### PR TITLE
[SPARK-36747][SQL] Do not collapse Project with Aggregate when correlated subqueries are present in the project list

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?
This PR adds a check in the optimizer rule `CollapseProject` to avoid combining Project with Aggregate when the project list contains one or more correlated scalar subqueries that reference the output of the aggregate. Combining Project with Aggregate can lead to an invalid plan after correlated subquery rewrite. This is because correlated scalar subqueries' references are used as join conditions, which cannot host aggregate expressions.

For example
```sql
select (select sum(c2) from t where c1 = cast(s as int)) from (select sum(c2) s from t)
```

```
== Optimized Logical Plan ==
Aggregate [sum(c2)#10L AS scalarsubquery(s)#11L] <--- Aggregate has neither grouping nor aggregate expressions.
+- Project [sum(c2)#10L]
   +- Join LeftOuter, (c1#2 = cast(sum(c2#3) as int))  <--- Aggregate expression in join condition
      :- LocalRelation [c2#3]
      +- Aggregate [c1#2], [sum(c2#3) AS sum(c2)#10L, c1#2]
         +- LocalRelation [c1#2, c2#3]

java.lang.UnsupportedOperationException: Cannot generate code for expression: sum(input[0, int, false])
```
Currently, we only allow a correlated scalar subquery in Aggregate if it is also in the grouping expressions.
https://github.com/apache/spark/blob/079a9c52925818532b57c9cec1ddd31be723885e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala#L661-L666

### Why are the changes needed?
To fix an existing optimizer issue.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit test.

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -982,6 +983,17 @@ object CollapseProject extends Rule[LogicalPlan] with AliasHelper {
...
   }
 
+  /**
+   * Check if the Project can be combined with the aggregate. If there exists a correlated
+   * scalar subquery in the project list that uses the output of the aggregate, the project
+   * cannot be collapsed. Otherwise the plan after subquery rewrite will not be valid.
+   */
+  private def canCollapseAggregate(p: Project, a: Aggregate): Boolean = {
+    p.projectList.forall(_.collect {
+      case s: ScalarSubquery if AttributeSet(s.children).intersect(a.outputSet).nonEmpty => s
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:993]: We should only check if `outerAttrs.nonEmpty`

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Comment by cloud-fan** [comment]
@allisonwang-db can you open a backport PR for 3.2?

**Pull request closed without merging**